source/fundamentals/data-formats/documents.txt

```diff
@@ -14,34 +14,351 @@ Documents
 
 BsonDocument
 ------------
 
-- Description
-- Link to Fundamentals > Document Data Formats: BSON for more information
-- Code example of declaring one and appending fields
-- Link to Fundamentals > CRUD for more information
-- Link to API documentation
+The ``BsonDocument`` class provides a type-safe API to access and manipulate
+a BSON document. You need to specify the BSON type from the Java BSON
+library for each field. See the following table for mappings between
+frequently-used BSON and Java BSON library types:
+
+.. list-table::
+   :header-rows: 1
+   :stub-columns: 1
+
+   * - BSON type
+     - Java BSON library type
+   * - Array
+     - ``org.bson.BsonArray``
+   * - Binary
+     - ``org.bson.BsonBinary``
+   * - Boolean
+     - ``org.bson.Boolean``
+   * - Date (long value)
+     - ``org.bson.BsonDateTime``
+   * - Document
+     - ``org.bson.BsonDocument``
+   * - Double
+     - ``org.bson.BsonDouble``
+   * - Int32
+     - ``org.bson.BsonInt32``
+   * - Int64
+     - ``org.bson.BsonInt64``
+   * - Null
+     - ``null``
+   * - ObjectId
+     - ``org.bson.BsonObjectId``
+   * - String
+     - ``org.bson.BsonString``
+
+
+You can also customize the type mapping by specifying a custom codec.
+For more information on customizing mapped types, see our guide on using
+:doc:`Codecs </fundamentals/data-formats/codecs>`.
+
+In the following code snippet, we show how to instantiate and build a sample
+``BsonDocument`` instance that contains a variety of data types:
+
+.. code-block:: java
+   :copyable: true
+
+   BsonDocument author = new BsonDocument()
+       .append("_id", new BsonObjectId())
+       .append("name", new BsonString("Gabriel García Márquez"))
+       .append("lastUpdated", new BsonDateTime(new Date(120, 2, 26).getTime()))
+       .append("novels", new BsonArray(Arrays.asList(
+           new BsonDocument().append("title", new BsonString("One Hundred Years of Solitude")).append("yearPublished", new BsonInt32(1967)),
+           new BsonDocument().append("title", new BsonString("Chronicle of a Death Foretold")).append("yearPublished", new BsonInt32(1981)),
+           new BsonDocument().append("title", new BsonString("Love in the Time of Cholera")).append("yearPublished", new BsonInt32(1985))
+      )));
+
+You can insert this document into a collection using the following code:
+
+.. code-block:: java
+   :copyable: true
+
+   // MongoClient mongoClient = ...
+
+   MongoCollection<BsonDocument> collection = database.getCollection("authors", BsonDocument.class);
+   MongoDatabase database = mongoClient.getDatabase("fundamentals_data");
+
+   InsertOneResult result = collection.insertOne(author);
+
+Once you perform a successful insert, you can retrieve the sample document
+data from the collection using the following code:
+
+.. code-block:: java
+   :copyable: true
+
+   import com.mongodb.client.model.Filters;
+   ...
+
+   BsonDocument doc = collection.find(Filters.eq("novels.yearPublished", 1967)).first();
```

Comment:
I wonder if this query could confuse users who aren't hugely familiar with MongoDB, since we're using some fairly advanced ideas here (querying an embedded document with dot syntax, matching a document based off a match to one field in a single document in a collection). On one hand, this does demonstrate some interesting MongoDB edge cases, and a lot of readers might be curious about dot syntax. On the other hand... when I started out with MongoDB, I might have had some trouble following this. Perhaps it would be better to use a simpler filter here and instead cover these more advanced concepts in the CRUD section?